<a href="https://colab.research.google.com/github/sasireddyvari/LifeCycle_DataScience/blob/master/FeatureEngineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**One hot encoding to handle multiple categories in one feature**

In [ ]:
!unzip '/content/drive/MyDrive/Colab Notebooks/mercedes-benz-data-exploration.zip'

Archive:  /content/drive/MyDrive/Colab Notebooks/mercedes-benz-data-exploration.zip
   creating: mercedes-benz-data-exploration/
   creating: mercedes-benz-data-exploration/sample_submission.csv/
  inflating: mercedes-benz-data-exploration/sample_submission.csv/sample_submission.csv  
   creating: mercedes-benz-data-exploration/test.csv/
  inflating: mercedes-benz-data-exploration/test.csv/test.csv  
   creating: mercedes-benz-data-exploration/train.csv/
  inflating: mercedes-benz-data-exploration/train.csv/train.csv  


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
data=pd.read_csv('/content/mercedes-benz-data-exploration/train.csv/train.csv',usecols=['X1','X2','X3','X4','X5','X6'])

In [ ]:
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [ ]:
for col in data:
  print(col ,':', len(data[col].unique()) , 'labels')

X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels


In [ ]:
data_encoding=pd.get_dummies(data,drop_first=True)

In [ ]:
data_encoding.shape

(4209, 117)

**As we did one hot encoding for 5 random variables but got 117 random variables. This might increase dimensionality , Hence one hot is done for most repeated categories**

In [ ]:
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [ ]:
len(data['X1'].unique())

27

In [ ]:
data['X1'].value_counts().sort_values(ascending=False)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
y      23
f      23
j      22
n      19
k      17
p       9
g       6
d       3
q       3
ab      3
Name: X1, dtype: int64

In [ ]:
data['X1'].value_counts().sort_values(ascending=False).head(20)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
y      23
f      23
j      22
Name: X1, dtype: int64

In [ ]:
data['X1'].value_counts().sort_values(ascending=False).head(10)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
Name: X1, dtype: int64

In [ ]:
top_10=[x for x in data.X1.value_counts().sort_values(ascending=False).head(10).index]
top_10

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [ ]:
def one_hot_top(df,variable,top_categories):
  for category in top_categories:
    data[variable +'_'+ category]=np.where(data[variable]==category,1,0)


(4209, 16)

In [ ]:
data=pd.read_csv('/content/mercedes-benz-data-exploration/train.csv/train.csv',usecols=['X1','X2','X3','X4','X5','X6'])

In [ ]:
for variable in data:
  one_hot_top(data,variable,top_10)
data.shape

(4209, 66)

**Count/Frequency Encoding : Ordinal Technique**

In [ ]:
data=pd.read_csv('/content/mercedes-benz-data-exploration/train.csv/train.csv',usecols=['X1','X2','X3','X4','X5','X6'])

In [ ]:
data.head(10)

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d
5  b   e  c  d  g  h
6  r   e  f  d  f  h
7  l  as  f  d  f  j
8  s  as  e  d  f  i
9  b  aq  c  d  f  a

In [ ]:
data.shape

(4209, 6)

In [ ]:
for i in data:
  print(i,len(data[i].unique()),'labels/categories')

X1 27 labels/categories
X2 44 labels/categories
X3 7 labels/categories
X4 4 labels/categories
X5 29 labels/categories
X6 12 labels/categories


In [ ]:
data.X1.value_counts().to_dict()

{'a': 143,
 'aa': 833,
 'ab': 3,
 'b': 592,
 'c': 121,
 'd': 3,
 'e': 33,
 'f': 23,
 'g': 6,
 'h': 29,
 'i': 203,
 'j': 22,
 'k': 17,
 'l': 590,
 'm': 32,
 'n': 19,
 'o': 82,
 'p': 9,
 'q': 3,
 'r': 251,
 's': 598,
 't': 31,
 'u': 37,
 'v': 408,
 'w': 52,
 'y': 23,
 'z': 46}

In [ ]:
data_frequency=data.X1.value_counts().to_dict()
data_frequency

{'a': 143,
 'aa': 833,
 'ab': 3,
 'b': 592,
 'c': 121,
 'd': 3,
 'e': 33,
 'f': 23,
 'g': 6,
 'h': 29,
 'i': 203,
 'j': 22,
 'k': 17,
 'l': 590,
 'm': 32,
 'n': 19,
 'o': 82,
 'p': 9,
 'q': 3,
 'r': 251,
 's': 598,
 't': 31,
 'u': 37,
 'v': 408,
 'w': 52,
 'y': 23,
 'z': 46}

In [ ]:
data.X1=data.X1.map(data_frequency)
print(data)

      X1  X2 X3 X4  X5 X6
0    NaN  at  a  d   u  j
1    NaN  av  e  d   y  l
2    NaN   n  c  d   x  j
3    NaN   n  f  d   x  l
4    NaN   n  f  d   h  d
...   ..  .. .. ..  .. ..
4204 NaN  as  c  d  aa  d
4205 NaN   t  d  d  aa  h
4206 NaN   r  a  d  aa  g
4207 NaN   e  f  d  aa  l
4208 NaN  ae  c  d  aa  g

[4209 rows x 6 columns]


In [ ]:
Disadvantage is Might work or might not work
The another one is if two categories repeated same, then we loose some data

**Ordinal Encoding**

In [70]:
import pandas as pd
import datetime

In [76]:
data=datetime.datetime.today()
data_list=[data-datetime.timedelta(days=x) for x in range(0,20)]
data_list

[datetime.datetime(2020, 12, 6, 15, 27, 5, 147457),
 datetime.datetime(2020, 12, 5, 15, 27, 5, 147457),
 datetime.datetime(2020, 12, 4, 15, 27, 5, 147457),
 datetime.datetime(2020, 12, 3, 15, 27, 5, 147457),
 datetime.datetime(2020, 12, 2, 15, 27, 5, 147457),
 datetime.datetime(2020, 12, 1, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 30, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 29, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 28, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 27, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 26, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 25, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 24, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 23, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 22, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 21, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 20, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 19, 15, 27, 5, 147457),
 datetime.datetime(2020, 11, 18, 15, 27, 5, 147457),

In [77]:
df = pd.DataFrame(data_list)

In [82]:
df.columns=['day']
df

day
0  2020-12-06 15:27:05.147457
1  2020-12-05 15:27:05.147457
2  2020-12-04 15:27:05.147457
3  2020-12-03 15:27:05.147457
4  2020-12-02 15:27:05.147457
5  2020-12-01 15:27:05.147457
6  2020-11-30 15:27:05.147457
7  2020-11-29 15:27:05.147457
8  2020-11-28 15:27:05.147457
9  2020-11-27 15:27:05.147457
10 2020-11-26 15:27:05.147457
11 2020-11-25 15:27:05.147457
12 2020-11-24 15:27:05.147457
13 2020-11-23 15:27:05.147457
14 2020-11-22 15:27:05.147457
15 2020-11-21 15:27:05.147457
16 2020-11-20 15:27:05.147457
17 2020-11-19 15:27:05.147457
18 2020-11-18 15:27:05.147457
19 2020-11-17 15:27:05.147457

In [84]:
df['dayofweek']=df['day'].dt.dayofweek

In [85]:
df

day  dayofweek
0  2020-12-06 15:27:05.147457          6
1  2020-12-05 15:27:05.147457          5
2  2020-12-04 15:27:05.147457          4
3  2020-12-03 15:27:05.147457          3
4  2020-12-02 15:27:05.147457          2
5  2020-12-01 15:27:05.147457          1
6  2020-11-30 15:27:05.147457          0
7  2020-11-29 15:27:05.147457          6
8  2020-11-28 15:27:05.147457          5
9  2020-11-27 15:27:05.147457          4
10 2020-11-26 15:27:05.147457          3
11 2020-11-25 15:27:05.147457          2
12 2020-11-24 15:27:05.147457          1
13 2020-11-23 15:27:05.147457          0
14 2020-11-22 15:27:05.147457          6
15 2020-11-21 15:27:05.147457          5
16 2020-11-20 15:27:05.147457          4
17 2020-11-19 15:27:05.147457          3
18 2020-11-18 15:27:05.147457          2
19 2020-11-17 15:27:05.147457          1

**Handling Missing Values:**

Missing completely at Random(MCAR): Absolutely No Relationship between the data missing and any other values

Missing Data Not at Random(MDAR): Some relationship

In [88]:
data=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/titanic.csv')
data.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [90]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [99]:
data[data['Embarked'].isnull()]

PassengerId  Survived  Pclass  ...  Fare Cabin  Embarked
61            62         1       1  ...  80.0   B28       NaN
829          830         1       1  ...  80.0   B28       NaN

[2 rows x 12 columns]

In [128]:
data['Cabin_null']=np.where(data['Cabin'].isnull(),1,0)
data.columns.get_loc('Cabin_null')

12

In [126]:
data.drop(['cabin_null'],inplace=True,axis=1)

In [129]:
data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Cabin_null'],
      dtype='object')

In [130]:
data['Cabin_null'].mean() # 77% of missing values

0.7710437710437711